File: apps/demo-fe-ts-tanstackstart/app/lib/api/tokens.ts
import { apiFetch } from "./client";
import type { JwksResponse } from "./types";

export function getJwks(): Promise<JwksResponse> {
  return apiFetch<JwksResponse>("/.well-known/jwks.json");
}

export function decodeTokenClaims(token: string): Record<string, unknown> {
  const parts = token.split(".");
  if (parts.length !== 3) throw new Error("Invalid JWT format");
  const payload = parts[1];
  if (!payload) throw new Error("Invalid JWT format");
  const decoded = atob(payload.replace(/-/g, "+").replace(/_/g, "/"));
  return JSON.parse(decoded) as Record<string, unknown>;
}

File: apps/demo-fe-elixir-phoenix/lib/demo_fe_exph/api/tokens.ex
defmodule DemoFeExph.Api.Tokens do
  @moduledoc "Token utilities: JWKS fetching and JWT claims decoding."

  alias DemoFeExph.Api.Client

  @doc "Fetches the JSON Web Key Set from the backend."
  def get_jwks do
    Client.get("/.well-known/jwks.json")
  end

  @doc """
  Decodes the claims from a JWT token without verification.

  Returns `{:ok, claims}` where claims is a map of the JWT payload,
  or `{:error, reason}` if the token is malformed.
  """
  def decode_token_claims(token) do
    with [_header, payload, _signature] <- String.split(token, "."),
         {:ok, decoded} <- Base.url_decode64(payload, padding: false),
         {:ok, claims} <- Jason.decode(decoded) do
      {:ok, claims}
    else
      _ -> {:error, :invalid_token}
    end
  end
end
